crates/ruma-events/src/call/member.rs

```diff
@@ -0,0 +1,386 @@
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Membership {
+    /// The type of the matrixRTC session the membership belongs to. (e.g. call, spacial,
+    /// document...)
+    #[serde(flatten)]
+    pub application: Application,
+    /// The device id of this membership. (The same user can join with their phone/computer)
+    pub device_id: String,
+    /// The duration in milliseconds relative to the time this membership joined
+    /// (`MIN(content.created_ts, event.origin_server_ts)`) during which the membership is valid.
+    pub expires: u64,
+    /// Contains the `origin_server_ts` of the initial session join.
+    /// If the membership is updated this field will be used to track to
+    /// original `origin_server_ts`
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub created_ts: Option<Timestamp>,
+    /// A list of the foci in use for this membership
+    pub foci_active: Vec<Foci>,
+    /// The id of the membership. This is required to guarantee uniqueness of the event.
+    /// (Sending the same state event twice to synapse makes the HS drop the second one and return
+    /// 200)
+    #[serde(rename = "membershipID")]
+    pub membership_id: String,
+}
+
+impl Membership {
+    /// Application is "m.call" and scope is "m.room"
+    pub fn is_room_call(&self) -> bool {
+        as_variant!(&self.application, Application::Call)
+            .is_some_and(|call| call.scope == CallScope::Room)
+    }
+
+    /// Application is "m.call"
+    pub fn is_call(&self) -> bool {
+        as_variant!(&self.application, Application::Call).is_some()
+    }
+
+    /// Check if the event is expired.
+    /// Defaults to using `created_ts` in the event content.
+    /// If no `origin_server_ts` is provided and the event does not contain `created_ts`
+    /// the event will be considered as not expired. (A warning will be logged)
+    ///
+    /// # Arguments
+    ///
+    /// * `origin_server_ts` - a fallback if `created_ts` is not present
+    pub fn is_expired(&self, origin_server_ts: Option<Timestamp>) -> bool {
+        let ev_created_ts = match (self.created_ts, origin_server_ts) {
+            (Some(created_ts), Some(_)) => Some(created_ts),
+            (None, Some(server_ts)) => Some(server_ts),
+            (Some(created_ts), None) => Some(created_ts),
+            _ => None,
+        };
+        if let Some(ev_created_ts) = ev_created_ts {
+            let now = Timestamp::now().to_system_time();
+            let expire_ts =
+                ev_created_ts.to_system_time().map(|t| t + Duration::from_millis(self.expires));
+            now > expire_ts
+        } else {
+            // This should not be reached since we only allow events that have copied over
+            // the origin server ts. `copy_origin_server_ts_to_membership`
+            warn!("Encountered a Call Member state event where the origin_ts (or origin_server_ts) could not be found.
```

Comment:
```suggestion
            warn!("Encountered a Call Member state event where the origin_ts (or origin_server_ts) could not be found.\
```
Otherwise the indentation at the beginning of the next line will appear in the logs.